firedrake/matrix_free/formmanipulation.py

```diff
@@ -0,0 +1,82 @@
+from __future__ import absolute_import
+
+import numpy
+
+from ufl import as_vector
+from ufl.classes import Zero
+from ufl.algorithms.map_integrands import map_integrand_dags
+from ufl.corealg.map_dag import MultiFunction
+
+from firedrake.ufl_expr import Argument
+
+
+class ExtractSubBlock(MultiFunction):
```

Comment:
I see, thanks. In that case I guess the original `FormSplitter` could be replaced with a for loop calling this one with appropriate arguments.